addons/sale_stock/models/sale_order.py

```diff
@@ -64,11 +65,32 @@ def _compute_expected_date(self):
                 expected_date = min(dates_list) if order.picking_policy == 'direct' else max(dates_list)
                 order.expected_date = fields.Datetime.to_string(expected_date)
 
+    @api.depends('user_id', 'company_id')
+    def _compute_warehouse_id(self):
+        for order in self:
+            company = order.company_id
+            oldOrder = order._origin
+            if oldOrder and oldOrder.company_id != order.company_id:
+                order.warehouse_id = False
+            if oldOrder and oldOrder.user_id != order.user_id:
+                order.warehouse_id = False
+            if not order.warehouse_id or not oldOrder:
+                if order.user_id:
+                    if order.user_id.with_company(company.id).property_warehouse_id:
+                        order.warehouse_id = order.user_id.with_company(company.id).property_warehouse_id.id
+            if not order.warehouse_id:
+                order.warehouse_id = order.env['stock.warehouse'].search([('company_id', '=', company.id)], limit=1)
```

Comment:
i don't see the point of making this field a stored computed, plus you're still using the default and the patch in create

while not keeping the old behaviour:
- default making a search on warehouse
- onchange
- patch create if the value is missing

but instead of these chunks getting the warehouse with the smallest sequence for the current company, they should call a new method: res.user._get_warehouse that will return either the property set or the warehouse with the smallest sequence

that would make a smaller and cleaner patch imo